### DISCUSSION TITLE
Pre-load files into flash for custom firmware.

### DISCUSSION THREAD

**Discussion by SomeSpaceNerd** [micropython/micropython#14144] [ESP32]
I was creating a custom firmware and need to load 3 files within a directory into flash memory as soon as the board gets flashed, is there a place to put this directory where it would get copied into flash, or do I have to do it afterwards by accessing the Filesystem? Also, I would like more clarification on where or how to put main.py into a custom firmware as well, as some of the documentation and discussions I found are outdated or not useful.

**Comment by walklinewang** [comment]
It seems to me that this is a port-related issue, some ports, such as `esp32` and `rp2`, they have their own `modules` folder for containing initialization files and/or user files compiled to firmware, while some ports don't.

**Comment by robert-hh** [comment]
If you build your own firmware, then you can use the manifest scheme to tell the place and names of the files to be included. There is one hiccup with `main.py` and `boot.py`. They cannot be frozen into the firmware and have to be loaded into the final board. 
You could however as a workaround create them in `_boot.py`. `_boot.py` is frozen in the firmware, as it usually mounts the file system. There you can add code to create a default `main.py` and eventually `boot.py` if they do not exist.

**Reply by SomeSpaceNerd** [reply]
My code that I would like to load is a OTA updater/bootloader, so replacing/appending _boot.py with it would be what I’m looking for. But the manifest system may not work for the types of files I need to load. Thank you for the answer!

**Reply by robert-hh** [reply]
There was a suggestion by @peterhinch to embed non-python files into a python script, which would be extracted on import of the script. Peter, do you still have the link?

**Reply by peterhinch** [reply]
It's [data-to-py](https://github.com/peterhinch/micropython_data_to_py). However freezefs (below) is a major superset.

Those interested in these capabilities may also be interested in [romfs](https://github.com/micropython/micropython/pull/8381). This is a line of development with huge potential.

**Comment by bixb922** [comment]
This may help for what you need:

https://github.com/bixb922/freezefs

Install with ```pip install freezefs```. Put all your files and subfolders in a folder called (for example) myfolder and run ```python -m freezefs  myfolder frozen_myfolder.py ```. This archives the files to ```frozen_myfolder.py``` and aooebds the needed driver. 

Add ```frozen_myfolder.py``` to the manifest.py and generate a Micropython image.

In boot.py or main.py add ```import frozen_myfolder```

The files now will appear as a readonly folder ```/myfolder``` as a virtual file system. 

(edit)
Other excellent and related options are:

https://github.com/micropython/micropython/pull/8191 (see status in the PR, it seems this will soon be in standard Micropython, and officially supported)
https://github.com/Josverl/vfs_merge 
https://github.com/peterhinch/micropython-samples/blob/master/data_to_py/data_to_py.py
https://github.com/glenn20/micropython-esp32-ota
https://github.com/glenn20/mp-image-tool-esp32

**Reply by Josverl** [reply]
Good collection of options, which I suggest would be good to add to the wiki

**Reply by andrewleech** [reply]
For reference I compiled some of these notes here: https://github.com/micropython/micropython/wiki/Bundling-and-deploying-application-code-and-files which is linked from the wiki tips and tricks page
